rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+    let header_blob = header_val.as_blob_mut();
+    let header_bytes: &mut [u8] =
+        core::slice::from_raw_parts_mut(header_blob.payload_addr(), header_len as usize);
+
+    // Move it:
+    read(0, header_bytes); // Save first block as "head block".
+    write(
+        (region0_blocks * (meta_data::size::BLOCK_IN_BYTES as u32)).into(),
+        header_bytes,
+    );
+
+    // Zero-out the temp blob.
+    for i in 0..header_bytes.len() {
+        header_bytes[i] = 0;
+    }
+    write(0, header_bytes); // Zero out first block, for region manager meta data.
+
+    /* Initialize meta data as if there is only region 0. */
+    meta_data::total_allocated_blocks::set(region0_blocks.into());
+    meta_data::total_allocated_regions::set(1);
+
+    /*
+     * Initialize region0's table entries, for "recovery".
+     */
+
+    /* head block is physically last, but logically first. */
+    let head_block_id = region0_blocks - 1; /* invariant: region0_blocks is non-zero. */
+
+    meta_data::block_region_table::set(BlockId(head_block_id as u16), Some((RegionId(0), 0)));
+    meta_data::region_table::set(&RegionId(0), Some(RegionSizeInPages(region0_pages as u64)));
+
+    /* Any other blocks that follow head block are numbered [0,...,head_block_id) */
+    /* They're logical placements are [1,...,) -- one more than their new identity, as a number. */
+    for i in 0..head_block_id {
+        meta_data::block_region_table::set(
+            BlockId((i) as u16),
+            Some((RegionId(0), (i + 1) as u16)),
+        );
+    }
+    /* "Recover" the region data into a heap object. */
+    REGION_0 = region_recover(mem, &RegionId(0));
+
+    // Ensure that regions 2 through 15 are already reserved for
+    // future use by future Motoko compiler-RTS features.
+    region_reserve_id_span(mem, Some(RegionId(1)), RegionId(15));
+}
+
+//
+// region manager migration/initialization, with pre-existing stable data.
+// Case: Version 2 into version 2 ("Trivial migration" case).
+//
+pub(crate) unsafe fn region_migration_from_v2_into_v2<M: Memory>(mem: &mut M) {
+    REGION_0 = region_recover(mem, &RegionId(0));
+
+    // Ensure that regions 2 through 15 are already reserved for
+    // future use by future Motoko compiler-RTS features.
+    region_reserve_id_span(mem, None, RegionId(15));
+}
+
+//
+// region manager migration/initialization, with pre-existing stable data.
+//
+// Valid cases:
+//  - from version 1 into version 2.
+//  - from version 2 into version 2.
+//
+#[ic_mem_fn]
+pub(crate) unsafe fn region_init<M: Memory>(mem: &mut M, from_version: i32) {
+    // Recall that we've done this later, without asking ic0_stable::size.
+    assert_eq!(crate::region::REGION_MEM_SIZE_INIT, false);
+    crate::region::REGION_MEM_SIZE_INIT = true;
+
+    match from_version {
+        0 => region_migration_from_v0_into_v2(mem),
+        2 => region_migration_from_v2_into_v2(mem),
+        1 => region_migration_from_v1_into_v2(mem),
```

Comment:
Little detail (only aesthetics): Potential symmetry:
```suggestion
        1 => region_migration_from_v1_into_v2(mem),
        2 => region_migration_from_v2_into_v2(mem),
```